### PR TITLE
feat: support openmldb spark connector(writer only)

### LINKED ISSUE #728: Support Spark connector to write data in online storage
Support Spark connector for online table.

### PR DESCRIPTION
**Pull request by vagetablechicken**
Close #728 
support that writing spark dataframe to openmldb online storage(using java client to write to tablet server).

### CODE REVIEW

File: java/openmldb-spark-connector/src/main/java/com/_4paradigm/openmldb/spark/write/OpenmldbBatchWrite.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    }
+
+    @Override
+    public DataWriterFactory createBatchWriterFactory(PhysicalWriteInfo info) {
+        return new OpenmldbDataWriterFactory(config);
+    }
+
+    @Override
+    public void commit(WriterCommitMessage[] messages) {
+
```
Comment (tobegit3hub) [inline java/openmldb-spark-connector/src/main/java/com/_4paradigm/openmldb/spark/write/OpenmldbBatchWrite.java:42]: Should we support batch write? Or throw an unsupported exception.

File: java/openmldb-spark-connector/src/main/java/com/_4paradigm/openmldb/spark/write/OpenmldbDataWriter.java [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+        } catch (SQLException e) {
+            e.printStackTrace();
+            throw new IOException(e.getMessage());
+        }
+    }
+
+    private void addRow(InternalRow record, PreparedStatement preparedStatement) throws SQLException {
+        // idx in preparedStatement starts from 1
+        ResultSetMetaData metaData = preparedStatement.getMetaData();
+        for (int i = 0; i < record.numFields(); i++) {
```
Comment (tobegit3hub) [inline java/openmldb-spark-connector/src/main/java/com/_4paradigm/openmldb/spark/write/OpenmldbDataWriter.java:84]: We may start with 1 for the for-loop.

File: java/openmldb-spark-connector/src/main/java/com/_4paradigm/openmldb/spark/write/OpenmldbDataWriter.java [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+            // sqlType
+            int type = metaData.getColumnType(i + 1);
+            switch (type) {
+                case Types.BOOLEAN:
+                    preparedStatement.setBoolean(i + 1, record.getBoolean(i));
```
Comment (tobegit3hub) [inline java/openmldb-spark-connector/src/main/java/com/_4paradigm/openmldb/spark/write/OpenmldbDataWriter.java:93]: Should we check if the value is null?

File: java/openmldb-spark-connector/src/main/java/com/_4paradigm/openmldb/spark/write/OpenmldbDataWriter.java [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+                    preparedStatement.setLong(i + 1, record.getLong(i));
+                    break;
+                case Types.FLOAT:
+                    preparedStatement.setFloat(i + 1, record.getFloat(i));
+                    break;
+                case Types.DOUBLE:
+                    preparedStatement.setDouble(i + 1, record.getDouble(i));
+                    break;
+                case Types.VARCHAR:
+                    preparedStatement.setString(i + 1, record.getString(i));
```
Comment (tobegit3hub) [inline java/openmldb-spark-connector/src/main/java/com/_4paradigm/openmldb/spark/write/OpenmldbDataWriter.java:111]: Should we escape the single quote or double quote?

Comment (vagetablechicken) [reply java/openmldb-spark-connector/src/main/java/com/_4paradigm/openmldb/spark/write/OpenmldbDataWriter.java:111]: We load file by DataFrameReader first, so DataFrameReader should do escape by add some options ,ref https://spark.apache.org/docs/3.2.0/sql-data-sources.html
I'll check how quote works in spark.

File: java/openmldb-spark-connector/src/main/java/com/_4paradigm/openmldb/spark/write/OpenmldbDataWriter.java (hunk 3 above)
Comment (vagetablechicken) [reply java/openmldb-spark-connector/src/main/java/com/_4paradigm/openmldb/spark/write/OpenmldbDataWriter.java:93]: Yes, I'll add null check.

File: java/openmldb-spark-connector/src/main/java/com/_4paradigm/openmldb/spark/write/OpenmldbDataWriter.java (hunk 2 above)
Comment (vagetablechicken) [reply java/openmldb-spark-connector/src/main/java/com/_4paradigm/openmldb/spark/write/OpenmldbDataWriter.java:84]: but in record, fields start from 0, actually it's a vector of columns.

File: java/openmldb-spark-connector/src/main/java/com/_4paradigm/openmldb/spark/write/OpenmldbBatchWrite.java (hunk 1 above)
Comment (vagetablechicken) [reply java/openmldb-spark-connector/src/main/java/com/_4paradigm/openmldb/spark/write/OpenmldbBatchWrite.java:42]: Data Source V2's write only support batch write/streaming write. Online storage can't support batch write(can't do transaction write), but we should use the batch write API of DataSourceV2.

Comment (tobegit3hub) [reply java/openmldb-spark-connector/src/main/java/com/_4paradigm/openmldb/spark/write/OpenmldbBatchWrite.java:42]: It seems that batch write is the basic function in jdbc. We may support batch write in spark once `openmldb-jdbc` supports it.

File: java/openmldb-spark-connector/src/main/java/com/_4paradigm/openmldb/spark/write/OpenmldbDataWriter.java (hunk 2 above)
Comment (tobegit3hub) [reply java/openmldb-spark-connector/src/main/java/com/_4paradigm/openmldb/spark/write/OpenmldbDataWriter.java:84]: Oh, did not see the usage of `record`.

File: java/openmldb-spark-connector/src/main/java/com/_4paradigm/openmldb/spark/write/OpenmldbDataWriter.java (hunk 4 above)
Comment (tobegit3hub) [reply java/openmldb-spark-connector/src/main/java/com/_4paradigm/openmldb/spark/write/OpenmldbDataWriter.java:111]: I'm worry about if our `preparedStatement` supports passing the string with double quote. Using string to generate SQL may have this issue but normally `preparedStatement` can handle this. If so, we do need to do extra work for string with double quote.

Comment (vagetablechicken) [reply java/openmldb-spark-connector/src/main/java/com/_4paradigm/openmldb/spark/write/OpenmldbDataWriter.java:111]: You mean a row like `ab“c, 1, 2` and quote is `"`, we 'll get a column `ab"c`(spark.read), tested ok.

**Review by tobegit3hub — APPROVED** [review]
LGTM

**Review by aceforeverd — APPROVED** [review]

**Comment by vagetablechicken** [comment]
[data import for offline and online in the task manager](https://trello.com/c/OqReyYcD/23-data-import-for-offline-and-online-in-the-task-manager)

**Comment by vagetablechicken** [comment]
[data import for offline and online in the task manager](https://trello.com/c/OqReyYcD/23-data-import-for-offline-and-online-in-the-task-manager)

**Pull request merged**